app/routes/_auth+/reset-password.tsx

```diff
@@ -16,29 +17,49 @@ import { GeneralErrorBoundary } from '~/components/error-boundary'
 import { authenticator, resetUserPassword } from '~/utils/auth.server'
 import {
 	Button,
+	ErrorList,
 	Field,
-	getFieldsFromSchema,
-	preprocessFormData,
-	useForm,
 } from '~/utils/forms'
+import { conform, useForm } from '@conform-to/react'
+import { getFieldsetConstraint, parse } from '@conform-to/zod'
 import { commitSession, getSession } from '~/utils/session.server'
 import { passwordSchema } from '~/utils/user-validation'
 import { resetPasswordSessionKey } from './forgot-password'
 
-const ResetPasswordSchema = z
-	.object({
-		password: passwordSchema,
-		confirmPassword: passwordSchema,
-	})
-	.superRefine(({ confirmPassword, password }, ctx) => {
-		if (confirmPassword !== password) {
-			ctx.addIssue({
-				path: ['confirmPassword'],
-				code: 'custom',
-				message: 'The passwords did not match',
-			})
-		}
-	})
+function createSchema(
+	constraints: {
+		doPasswordsMatch: (password: string, confirmPassword: string) => boolean;
+	} = {
+			doPasswordsMatch: () => true,
+		},
+) {
+	const ResetPasswordSchema = z
+		.object({
+			password: passwordSchema,
+			confirmPassword: passwordSchema,
+		})
+		.superRefine(({ confirmPassword, password }, ctx) => {
+			if (typeof constraints.doPasswordsMatch === 'undefined') {
+				// Validate only if the constraint is defined
+				ctx.addIssue({
+					code: z.ZodIssueCode.custom,
+					message: conform.VALIDATION_UNDEFINED,
+				});
+			} else {
+				const match = constraints.doPasswordsMatch(password, confirmPassword)
+				if (match) {
+					return;
+				}
+
+				ctx.addIssue({
+					path: ['confirmPassword'],
+					code: 'custom',
+					message: 'The passwords did not match',
+				})
+			}
+		})
```

Comment:
This pattern is needed only if you want to do async validation. A simple refine should be enough in this case.

```ts
.refine(({ confirmPassword, password }) => password === confirmPassword, {
    message: 'The passwords did not match',
    path: ['confirmPassword'],
});
```